feat(root) implement external signer for mpcv2

alebusse opened the pull request:
added base utils for MPCv2 external signer
added MPCv2 base code to support external signer
Implemented singing using external signer for MPCv2
Use the hashed signableTx as Adata to verify that the encrypted session matches with the tx during
the second and third DSG rounds in external signer, this would prevent re using the stored session
for a different tx

WP-2136
<!--
# Please be aware of the following when making your pull request:

## Description

Please include a summary of your proposed changes and which issue is being addressed. Please also include relevant motivation and context. List any dependencies that are required for this change.

## Issue Number

Internal Users - Please include the related internal tracking number (e.g. BG-000000).
External Users - Please link to any relevant github issues as necessary.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration.

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] My code compiles correctly for both Node and Browser environments
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My commits follow [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) and I have properly described any BREAKING CHANGES
- [ ] The ticket or github issue was included in the commit message as a reference
- [ ] I have made corresponding changes to the documentation and on any new/updated functions and/or methods - [jsdoc](https://jsdoc.app/)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
-->

alebusse committed:
feat(sdk-core): add base utils for MPCv2 external signer

added base utils for MPCv2 external signer

WP-2136

TICKET: WP-2136

alebusse committed:
feat(express): add external signer base code for MPCv2

added MPCv2 base code to support external signer

WP-2136

TICKET: WP-2136

alebusse committed:
feat(sdk-core): implemented signing with external signer for MPCv2

Implemented singing using external signer for MPCv2

WP-2136

TICKET: WP-2136

zahin-mohammad commented on modules/express/src/clientRoutes.ts:
```diff
@@ -828,7 +855,7 @@ function createSendParams(req: express.Request) {
 }
 
-function createTSSSendParams(req: express.Request) {
+function createTSSSendParams(req: express.Request, wallet?: Wallet) {
```
shouldn't wallet be required? Are there routes that aren't passing it in?

alebusse replied:
its not required for Eddsa since we dont check the multisigTypeVersion, but i can make it required

zahin-mohammad commented on modules/sdk-core/src/bitgo/utils/tss/baseTypes.ts:
```diff
@@ -107,6 +110,39 @@ export interface CustomGShareGeneratingFunction {
   }): Promise<GShare>;
 }
+
+export interface CustomMPCv2Round1GeneratingFunction {
```
Nit: signing and key gen both have multiple rounds, would be clearer if we specified signing here.
```suggestion
export interface CustomMPCv2SigningRound1GeneratingFunction {
```

zahin-mohammad commented on modules/sdk-core/src/bitgo/utils/tss/ecdsa/ecdsaMPCv2.ts:
```diff
@@ -724,5 +718,306 @@ export class EcdsaMPCv2Utils extends BaseEcdsaUtils {
+    const { txRequest, reqId } = params;
+    let txRequestResolved: TxRequest;
+
+    assert(
+      requestType === RequestType.tx,
+      'Only transaction signing is supported for external signer, got: ' + requestType
+    );
```
Lets add a ticket to support message signing with external signer. Don't want to let it fall through the cracks even though it is not needed atm.

zahin-mohammad commented on modules/sdk-core/src/bitgo/wallet/wallet.ts:
```diff
@@ -3152,6 +3161,54 @@ export class Wallet implements IWallet {
+      const signedTxRequest = await this.tssUtils.signEcdsaMPCv2TssUsingExternalSigner(
+        {
+          txRequest: txRequestId,
+          prv: '',
```
Nit: we should allow undefined `prv` if the custom signing functions are passed. Seems weird to pass an empty string.

alebusse replied:
Ya, its mostly because its using the wrong type. i will fix it

zahin-mohammad commented on modules/sdk-core/src/bitgo/utils/tss/baseTSSUtils.ts:
```diff
@@ -154,6 +157,26 @@ export default class BaseTssUtils<KeyShare> extends MpcUtils implements ITssUtil
+    externalSignerMPCv2Round3Generator: CustomMPCv2Round3GeneratingFunction,
+    requestType?: RequestType
+  ): Promise<TxRequest> {
+    throw new Error('Method not implemented.');
```
This still needs to be implemented in this PR right? Or will this be addressed in a different PR?

alebusse replied:
Nop, the base utils use the non implemented errors incase you used the wrong util (like eddsa)

alebusse committed:
test(express): add external signer test for MPCv2

added external signer test for MPCv2 routes

TICKET: WP-2136

alebusse committed:
feat(root): add Adata validation for DSG session validation

Use the hashed signableTx as Adata to verify that the encrypted session matches with the tx during
the second and third DSG rounds in external signer, this would prevent re using the stored session
for a different tx

WP-2136

TICKET: WP-2136
